server/src/main/java/org/elasticsearch/cluster/metadata/MetadataUpdateSettingsService.java

```diff
@@ -87,183 +98,218 @@ public MetadataUpdateSettingsService(
-                        indicesService.verifyIndexMetadata(updatedMetadata, updatedMetadata);
-                    }
-                } catch (IOException ex) {
-                    throw ExceptionsHelper.convertToElastic(ex);
-                }
+            if (changed == false) {
+                return currentState;
+            }
+
+            ClusterState updatedState = ClusterState.builder(currentState)
+                .metadata(metadataBuilder)
+                .routingTable(routingTableBuilder == null ? currentState.routingTable() : routingTableBuilder.build())
+                .blocks(changedBlocks ? blocks.build() : currentState.blocks())
+                .build();
 
-                return updatedState;
+            try {
+                for (Index index : openIndices) {
+                    final IndexMetadata currentMetadata = currentState.metadata().getIndexSafe(index);
+                    final IndexMetadata updatedMetadata = updatedState.metadata().getIndexSafe(index);
+                    indicesService.verifyIndexMetadata(currentMetadata, updatedMetadata);
+                }
+                for (Index index : closedIndices) {
+                    final IndexMetadata currentMetadata = currentState.metadata().getIndexSafe(index);
+                    final IndexMetadata updatedMetadata = updatedState.metadata().getIndexSafe(index);
+                    // Verifies that the current index settings can be updated with the updated dynamic settings.
+                    indicesService.verifyIndexMetadata(currentMetadata, updatedMetadata);
+                    // Now check that we can create the index with the updated settings (dynamic and non-dynamic).
+                    // This step is mandatory since we allow to update non-dynamic settings on closed indices.
+                    indicesService.verifyIndexMetadata(updatedMetadata, updatedMetadata);
+                }
+            } catch (IOException ex) {
+                throw ExceptionsHelper.convertToElastic(ex);
             }
-        };
 
-        clusterService.submitStateUpdateTask("update-settings " + Arrays.toString(request.indices()), clusterTask, this.executor);
+            return updatedState;
+        }
+    }
+
+    record MetadataUpdateSettingsTask(UpdateSettingsClusterStateUpdateRequest request, ActionListener<AcknowledgedResponse> listener)
+        implements
+            ClusterStateTaskListener,
+            ClusterStateAckListener {
+
+        @Override
+        public void onFailure(Exception e) {
+            listener.onFailure(e);
+        }
+
+        @Override
+        public boolean mustAck(DiscoveryNode discoveryNode) {
+            return true;
+        }
+
+        @Override
+        public void onAllNodesAcked() {
+            listener.onResponse(AcknowledgedResponse.of(true));
+        }
+
+        @Override
+        public void onAckFailure(Exception e) {
+            listener.onResponse(AcknowledgedResponse.of(false));
```

Comment:
Just realised we have constants for these things:

```suggestion
            listener.onResponse(AcknowledgedResponse.FALSE);
```